content/15.development_procedure.md

```diff
@@ -92,15 +92,26 @@ By queuing up pull requests for merging, it allows maintainers to mark a set of
-In addition to code formatting, other automated methods for code consistency are applied; at present, this includes ensuring that all ``import`` statements are sorted (according to alphabetical order within a first/second/third-party categorization).
-Past iterations of this automated formatting have also included ensuring that ["f-strings"](https://docs.python.org/3/tutorial/inputoutput.html#formatted-string-literals) are used wherever possible, but at present we have slightly loosened this requirement.
 
-For the purposes of checking for common code mistakes, we apply [``flake8``](https://flake8.pycqa.org/en/latest/), which reports unused variables, undefined variables, and other style violations that cannot be automatically corrected.
+We rely on the [``pre-commit``](https://pre-commit.com) framework, which enables automated checks as well as automatic fixes at commit time.
+This tool is an opt-in so not every "drive-by" contributor has to learn and install it, but continuous linting is provided by [pre-commit.ci](https://pre-commit.ci) so styling errors cannot slip in.
 
-All changes that can be applied via automation (specifically, code formatting) are accessible from within the GitHub pull request interface.
-This allows "drive-by" contributions to have their pull requests updated inline by an automated process, reducing the need to manually install packages to apply the changes.
+We configure `pre-commit` to run a mixed bag of formatters and static checkers.
+The former modify the code in place, while the latter only report errors and code smells (such as undefined variables, unused imports, bare `except` statements...) but let human developers fix them.
+
+Our suite of formatter most proeminently includes
+[``black``](https://black.readthedocs.io/en/stable/),
+[``isort``](https://pycqa.github.io/isort/) and
+[``pyupgrade``](https://github.com/asottile/pyupgrade).
+``black`` has been designed to maximize readability with as few free parameters as possible
+(In many ways, the fact that most of the ``yt`` code developers did not utilize this style before it was enforced likely enabled its uptake, as it was seen as a choice that "made everyone compromise."),
+while ``isort`` ensures that all ``import`` statements are sorted (according to alphabetical order within a first/second/third-party categorization), and ``pyupgrade`` modernizes some Python idioms according to our minimal support version of the language.
+In particular, ``pyupgrade`` enforces the usage of modern Python ["f-strings"](https://docs.python.org/3/tutorial/inputoutput.html#formatted-string-literals) since we do not support Python version older than 3.6.
+
+For static code analysis we rely on the [``flake8``](https://flake8.pycqa.org/en/latest/) framework (an possibly mypy for the specialized task of type checking ??)
```

Comment:
```suggestion
For static code analysis we rely on the [``flake8``](https://flake8.pycqa.org/en/latest/) framework.
Effort is underway to enable using `mypy` for the specialized task of type checking, but this is far from completion.
```

Reply:
actually not that far. Merging https://github.com/yt-project/yt/pull/3546 would suffice making yt "type-checked". Of course there's a lot more that we can do from there to _improve typing_, but as long as we are able to run mypy without error, we're "type-checked" !

Reply:
I agree